exercises/symbolic_sequences_data/descr.md

```diff
@@ -0,0 +1,177 @@
+`empty`,
+`singleton`,
+`sum`,
+`product`,
+and `map`.
+
+The constant `empty` is the empty sequence.
+
+```
+  val empty: 'a seq
+```
+
+The sequence `singleton x` has length 1. Its single element is `x`.
+
+```
+  val singleton: 'a -> 'a seq
+```
+
+The sequence `sum s1 s2` is the concatenation of the sequences `s1` and `s2`.
+
+```
+  val sum: 'a seq -> 'a seq -> 'a seq
+```
+
+The sequence `product s1 s2`, a sequence of pairs, is the Cartesian product
+of the sequences `s1` and `s2`. Its length is the product of the lengths of
+`s1` and `s2`. The first pair component is considered most significant,
+which means that the product of the integer sequence `0; 1` by itself is the
+sequence `(0, 0); (0, 1); (1, 0); (1, 1)`.
+
+```
+  val product: 'a seq -> 'b seq -> ('a * 'b) seq
+```
+
+The sequence `map phi s` is the image of the sequence `s` through the function
+`phi`.
+
+```
+  val map: ('a -> 'b) -> 'a seq -> 'b seq
+```
+
+### Destructors
+
+There are three destructors: `length`, `get`, and `foreach`.
+
+`length s` returns the length of the sequence `s`.
+
+```
+  val length: 'a seq -> int
+```
+
+If the index `i` is comprised between zero (included) and `length s` (excluded),
+then `get s i` returns the `i`-th element of the sequence `s`.
+Otherwise, `get s i` raises the exception `OutOfBounds`.
+
+```
+  exception OutOfBounds
+  val get: 'a seq -> int -> 'a
+```
+
+`foreach s k` iterates over all elements of the sequence `s`.
+Each element in turn is passed to the loop body `k`.
+Thus, the OCaml code `foreach s (fun x -> ...)` can be
+informally understood as `for each x in s do ...`.
+
+```
+  val foreach: 'a seq -> ('a -> unit) -> unit
+```
+
+**Note.**
+Because sequences can be extremely long,
+OCaml's built-in type `int`, which is 63 bits wide at best,
+can be insufficient to represent the length of a sequence.
+In order to avoid overflows, we *should* use a type of unbounded integers,
+such as the type `Z.t` provided by the library `zarith`.
+However, for the sake of simplicity,
+in this exercise, we stick with the type `int`.
+
+## Implementation
+
+In this exercise, we represent a sequence in memory as a value of the
+following algebraic data type:
+
+```
+type _ seq =
+| Empty    : 'a seq
+| Singleton: 'a -> 'a seq
+| Sum      : int * 'a seq * 'a seq -> 'a seq
+| Product  : int * 'a seq * 'b seq -> ('a * 'b) seq
+| Map      : int * ('a -> 'b) * 'a seq -> 'b seq
+```
+
+This is a *generalized* algebraic data type.
+In an ordinary algebraic data type,
+every data constructor has result type `'a seq`.
+Here, the constructor `Product` violates this rule:
+its result type is `('a * 'b) seq`.
+
+The data constructors `Sum`, `Product` and `Map` carry
+the length of the sequence, which has type `int`,
+as their first argument.
+This allows a constant-time implementation of the function `length`.
+
+## Assignment
+
+**Question 1.** Implement `length`.
+
+**Question 2.** Implement the five constructors presented above,
```

Comment:
"five constructors" -> "five constructor functions" (to avoid the ambiguity with the *data* constructors of the GADT). It may be even better to explicitly list them.

Reply:
Done.